07-read-write-plot.Rmd

```diff
@@ -136,6 +136,62 @@ world2 = spData::world
 world3 = st_read(system.file("shapes/world.gpkg", package = "spData"))
 ```
 
+## Geographic web services
+
+In an effort to standardize web APIs for accessing spatial data, the Open Geospatial Consortium (OGC) has created a number of specifications for web services (collectively known as OWS, which is short for OGC Web Services).
+These specifications include the Web Feature Service (WFS), Web Map Service (WMS), Web Map Tile Service (WMTS), the Web Coverage Service (WCS) and even a Wep Processing Service (WPS).
+Map servers such as PostGIS have adopted these protocals, leading to standardization of  queries:
+Like other web APIs, OWS APIs use a 'base URL' and a 'query string' proceeding a `?` to request data.
+
+There are many requests that can be made to a OWS service.
+One of the most fundamental is `getCapabilities`, demonstrated with the **httr** package to show how API queries can be constructed and dispatched, in this case to discover the capabilities of a service providing Dutch open data:
+
+```{r, eval=FALSE}
+base_url = "http://geodata.nationaalgeoregister.nl/bag/wfs"
+q = list(request = "GetCapabilities")
+res = httr::GET(url = base_url, query = q)
+res$url
+```
+
+The above code chunk demonstrates how API requests can be constructed programmatically with the `GET()` function, which takes a base URL and a list of query parameters which can easily be extended.
+The result of the request is saved in the object `res`, an object of class `response` defined in the **httr** package, which is a list containing information of the request, including the URL.
+As can be seen by executing `browseURL(res$url)`), the results can also be read directly in a browser, although they are not designed to be human readable.
+The contents of the request can be extracted as follows:
+
+```{r, eval=FALSE}
+txt = httr::content(res, "text")
+xml2::read_xml(txt)
+```
+
+```{r}
+#> {xml_document} ...
+#> [1] <ows:ServiceIdentification>\n  <ows:Title>BAG</o ...
+#> [2] <ows:ServiceProvider>\n  <ows:ProviderName>PDOK< ...
+#> [3] <ows:OperationsMetadata>\n  <ows:Operation name= ...
+#> [4] <FeatureTypeList>\n  <FeatureType>\n    <Name>ba ...
+#> [5] <fes:Filter_Capabilities>\n  <fes:Conformance>\n ...
+```
+
+The previous example shows how to gain low-level access to web services using **httr**, which can be useful for understanding how web services work.
+For many everyday tasks, however, a higher-level interface may be more appropriate, and a number of R packages, and tutorials, have been developed precisely for this purpose, as outlined below (an in-depth coverage of OWS services is beyond the scope of the book).
+
+Packages **ows4R**, **rwfs** and **sos4R** have been developed for working with OWS services in general, WFS and the sensor observation service (SOS) respectively.
+As of October 2018 all of them are still works in-progress and only one (**ows4R**) is on CRAN.
```

Comment:
Heads-up @eblondel we've mentioned your pkg in our open source book! Please take a look at mention of it and let me know if you have comments about our description of **ows4r** (looks like an interesting project) or the section in general.

Reply:
Thanks @Robinlovelace for mentioning ows4R in your book, indeed it's great!

I want to be very clear first on some aspect: ``ows4R`` is a new package, and giving its scope (OGC web-services), you can assume this will be "in development" for a lot of time :-), because it is expected to grow, and grow more.  This doesn't prevent that ows4R is going into release lifecycle with CRAN, and this means that stable software is expected to be released (of course we can still find bugs, it's quite natural). If we stop dev on it, this would be rather a bad news, and backward compability will be ensured unless a major enhancement would need user code adapation.

For your book, I would suggest to identifies what does OWS standards for, earlier in the text (and if not yet done.), ie OGC Web-Services, (with a link the OGC standards web-page, maybe?)

Maybe you could also mention the support of the Catalogue Service for the Web (CSW) standard, which targets metadata management, with binding with the [geometa](https://github.com/eblondel/geometa) package, for writing and reading of ISO/OGC geographic metadata.  FYI, these 2 packages are under active development. For what concerns ``ows4R``, any OGC standard service is potentially candidate for integration into ows4R, with priority given to standards that are not supported at all yet in R (while some other standard web-service have some implementations available like sos4R). A prototype for WCS (Web Coverage Service) is under investigation with mapping to the raster package, not that easy, WCS is much more complex that the simple WFS, this will target data accessing in multi-dimensional geographic data. I hope that a first client could be released next year.

If you want more information, i'm available.

Reply:
Many thanks for the detailed comment!

Reply:
To response to some of your suggestions, we do state what OWS early so that's good and link to the page and other resources later on. We could go into more detail. But the most important thing in this section I think, is that it exists! What I mean by that is that in a chapter on data I/O it's worth flagging all these (awesome) standard services. Because the R-OWS interfaces do not seem to be advanced the main priority is to point people in the right direction in a future proof way.

It sounds like `owsR` is stable and future-proof. Would it be possible to use it to download some features from a web service, building on the following reproducible example?

``` r
base_url = "http://geodata.nationaalgeoregister.nl/bag/wfs"
q = list(request = "GetCapabilities")
res = httr::GET(url = base_url, query = q)
res$url
#> [1] "http://geodata.nationaalgeoregister.nl/bag/wfs?request=GetCapabilities"
txt = httr::content(res, "text")
#> No encoding supplied: defaulting to UTF-8.
xml2::read_xml(txt)
#> {xml_document}
#> <WFS_Capabilities version="2.0.0" schemaLocation="http://www.opengis.net/wfs/2.0 http://schemas.opengis.net/wfs/2.0/wfs.xsd http://inspire.ec.europa.eu/schemas/inspire_dls/1.0 http://inspire.ec.europa.eu/schemas/inspire_dls/1.0/inspire_dls.xsd" updateSequence="1913" xmlns:wfs="http://www.opengis.net/wfs/2.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.opengis.net/wfs/2.0" xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:gml="http://www.opengis.net/gml/3.2" xmlns:fes="http://www.opengis.net/fes/2.0" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:inspire_dls="http://inspire.ec.europa.eu/schemas/inspire_dls/1.0" xmlns:inspire_common="http://inspire.ec.europa.eu/schemas/common/1.0" xmlns:bag="http://bag.geonovum.nl">
#> [1] <ows:ServiceIdentification>\n  <ows:Title>BAG</ows:Title>\n  <ows:Ab ...
#> [2] <ows:ServiceProvider>\n  <ows:ProviderName>PDOK</ows:ProviderName>\n ...
#> [3] <ows:OperationsMetadata>\n  <ows:Operation name="GetCapabilities">\n ...
#> [4] <FeatureTypeList>\n  <FeatureType>\n    <Name>bag:ligplaats</Name>\n ...
#> [5] <fes:Filter_Capabilities>\n  <fes:Conformance>\n    <fes:Constraint  ...
```

<sup>Created on 2018-10-18 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

Or at least point us in the right direction (will look at your other comment also, very informative). Note: we don't want to make this section too long!